Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,199 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A base ByteBuffer "Slicer", this base one just splits by a separator.
+public struct ByteBufferSliceSplitIterator: IteratorProtocol {
+    
+    let byteBuffer: ByteBuffer
+    let separator: UInt8
+    let start: Int
+    let length: Int
+    
+    private var currentIndex: Int
+    
+    public typealias Element = ByteBuffer
+    mutating public func next() -> ByteBuffer? {
+        
+        let slicingParameters = byteBuffer.withVeryUnsafeBytes { pointer -> (start: Int, length: Int) in
+            
+            let allBuffer = pointer.bindMemory(to: UInt8.self)
+                .dropFirst(self.start)
+                .prefix(length)
+            
+            let remaining = allBuffer.dropFirst(self.currentIndex)
```

Comment:
You can simplify some of this maths by changing `currentIndex` to be the index into the `ByteBuffer` you want to start from. This index is relative to the start of the buffer (and so the start of the pointer that was passed in), which allows you to simply `dropFirst(self.currentIndex)` to correctly set the start of the buffer pointer.

Reply:
This would also let you remove `start` by changing `length` to `endIndex` instead.